state.go

```diff
@@ -0,0 +1,514 @@
+	}
+	valueOp := iavl.NewValueOp(key, proof)
+	return doc, valueOp, nil
+}
+
+func (state *KatzenmintState) isAuthorized(addr string) bool {
+	if _, ok := state.validators[addr]; ok {
+		return false
+	}
+	return true
+}
+
+func (state *KatzenmintState) GetAuthorized(addr string) (pc.PublicKey, bool) {
+	pubkey, ok := state.validators[addr]
+	return pubkey, ok
+}
+
+func (state *KatzenmintState) isDescriptorAuthorized(desc *pki.MixDescriptor) bool {
+	pk := desc.IdentityKey.ByteArray()
+
+	switch desc.Layer {
+	// case 0:
+	// 	return state.authorizedMixes[pk]
+	case pki.LayerProvider:
+		// check authorities, should use validator address
+		_, ok := state.validators[bytesToAddress(pk)]
+		return ok
+	default:
+		return false
+	}
+}
+
+func (state *KatzenmintState) isDocumentAuthorized(doc *pki.Document) bool {
+	if _, ok := state.documents[doc.Epoch]; ok {
+		return true
+	}
+	return true
+}
+
+func (state *KatzenmintState) Set(key []byte, value []byte) error {
+	state.tree.Set(key, value)
+	return nil
+}
+
+func (state *KatzenmintState) Delete(key []byte) error {
+	_, success := state.tree.Remove(key)
+	if !success {
+		return fmt.Errorf("remove from database failed")
+	}
+	return nil
+}
+
+func (state *KatzenmintState) Get(key []byte) ([]byte, error) {
+	_, val := state.tree.Get(key)
+	if val == nil {
+		return nil, fmt.Errorf("key '%v' does not exist", key)
+	}
+	ret := make([]byte, len(val))
+	copy(ret, val)
+	return ret, nil
+}
+
+// Note: Caller ensures that the epoch is the current epoch +- 1.
+func (state *KatzenmintState) updateMixDescriptor(rawDesc []byte, desc *pki.MixDescriptor, epoch uint64) (err error) {
+	state.Lock()
+	defer state.Unlock()
+
+	pk := desc.IdentityKey.ByteArray()
+
+	// Get the public key -> descriptor map for the epoch.
+	m, ok := state.descriptors[epoch]
+	if !ok {
+		m = make(map[[eddsa.PublicKeySize]byte]*descriptor)
+		state.descriptors[epoch] = m
+	}
+
+	// Check for redundant uploads.
+	if d, ok := m[pk]; ok {
+		if d.raw == nil {
+			return fmt.Errorf("state: Wtf, raw field of descriptor for epoch %v is nil", epoch)
+		}
+		// If the descriptor changes, then it will be rejected to prevent
+		// nodes from reneging on uploads.
+		if !bytes.Equal(d.raw, rawDesc) {
+			return fmt.Errorf("state: Node %v: Conflicting descriptor for epoch %v", desc.IdentityKey, epoch)
+		}
+
+		// Redundant uploads that don't change are harmless.
+		return nil
+	}
+
+	// Ok, this is a new descriptor.
+	if state.documents[epoch] != nil {
+		// If there is a document already, the descriptor is late, and will
+		// never appear in a document, so reject it.
+		return fmt.Errorf("state: Node %v: Late descriptor upload for for epoch %v", desc.IdentityKey, epoch)
+	}
+
+	// Persist the raw descriptor to disk.
+	key := storageKey(descriptorsBucket, desc.IdentityKey.Bytes(), epoch)
+	if err := state.Set(key, rawDesc); err != nil {
+		return err
+	}
+
+	// Store the raw descriptor and the parsed struct.
```

Comment:
Should the state change be applied directly, given that it could be reverted when a transaction failed? Also, this makes the above check for redundant uploads actually take effect.